### PR TITLE
feat: support for soft deleted aspects

### PR DESCRIPTION
**Pull request by jywadhwani**
## Checklist
This change adds support for soft deleted aspects. Primarily it does the following
1. Makes the `metadata` column in `metadata_aspect` table nullable. The field will be `null` for a soft deleted aspect
2. Adds a method `delete` in `BaseLocalDAO` to soft delete an aspect. Note that the old `add` method stays the same i.e. we only support adding metadata that is _not null_ in `add` method. Moved the code common to `add` and `delete` method in a separate function.

This PR does _not_ cover the following
1. Sending MAE with appropriate values for soft deleted aspect
2. Changes to local secondary index for soft deleted aspect

- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [X] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 1]
```diff
@@ -364,12 +374,21 @@ public boolean isLocalSecondaryIndexEnabled() {
   /**
    * Similar to {@link #add(Urn, Class, Function, AuditStamp)} but takes the new value directly.
    */
-  @Nonnull
+  @Nullable
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:377]: Returning a `null` value made sense to me, because the metadata being inserted is actually `null`. Also returning an `Optional` value looks more of a breaking change, since the clients/GMS calling this method directly will now have to do `<return>.get()` to process the value.
What are your thoughts, particularly on minimizing the impact of this change on OSS users?

[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:377]: Makes sense. Returning an `Optional` value will prevent null pointer exceptions

File: dao-impl/ebean-dao/src/main/resources/gma-create-all.sql [hunk 2]
```diff
@@ -2,7 +2,7 @@ create table metadata_aspect (
   urn                           varchar(500) not null,
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-impl/ebean-dao/src/main/resources/gma-create-all.sql:2]: I would do that in a separate PR to limit the scope of this one. We can add the motivation as well for increasing the size - it isn't quite aligned with soft deleted aspects requirement.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java [hunk 3]
```diff
@@ -70,9 +70,8 @@
   @Index
   protected PrimaryKey key;
 
-  @NonNull
   @Lob
-  @Column(name = METADATA_COLUMN, nullable = false)
+  @Column(name = METADATA_COLUMN)
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java:74]: I don't think we need to specify `nullable=true` since the default value is `true`
https://docs.jboss.org/hibernate/jpa/2.1/api/javax/persistence/Column.html

File: restli-resources/src/main/java/com/linkedin/metadata/restli/BaseEntityResource.java [hunk 4]
```diff
@@ -197,6 +197,22 @@ protected BaseRestliAuditor getAuditor() {
     });
   }
 
+  /**
+   * An action method for soft deleting an aspect.
+   */
+  @Action(name = ACTION_SOFT_DELETE_ASPECT)
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply restli-resources/src/main/java/com/linkedin/metadata/restli/BaseEntityResource.java:203]: I already have a PR for `@RestMethod.Delete` https://github.com/linkedin/datahub-gma/pull/133
I added this method in `BaseEntityResource` similar to `ACTION_INGEST`.
Let me remove this change from the PR for now. I will create a separate PR on the resource level changes, to keep the scope of this one limited.

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/ListResultMetadata.pdl [hunk 5]
```diff
@@ -1,35 +1,17 @@
 namespace com.linkedin.metadata.query
 
-import com.linkedin.common.AuditStamp
-import com.linkedin.common.Urn
-
 /**
  * The additional metadata for a list result
  */
 record ListResultMetadata {
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/ListResultMetadata.pdl:6]: So `ListResult` returns a list of non-null items and so soft-deleted aspects won't be captured in this list. `ListResult` has `ListResultMetadata` that currently contains audit history. My original thought of adding soft-deleted aspects metadata to `ListResultMetadata` was to provide the history of soft-deleted aspects somewhere. But thinking about it again, it looks like over-engineering at this point.

If we do see a use of returning soft-deleted aspect metadata in `ListResult`, then we can consider this approach or another better one, if it exists.

Thanks for the review.

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/ListResultMetadata.pdl [hunk 6]
```diff
@@ -1,35 +1,17 @@
...
-      /**
-       * The urn for the entity that owns the metadata aspect
-       */
-      urn: Urn
-
-      /**
-       * The version number of the metadata aspect.
-       */
-      version: long
+  softDeletedAspects: array[ExtraInfo]
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/ListResultMetadata.pdl]: Removing this field.

File: dao-api/src/main/java/com/linkedin/metadata/dao/AspectWithExtraInfo.java [hunk 7]
```diff
@@ -12,7 +13,7 @@
 @Value
 public class AspectWithExtraInfo<ASPECT extends RecordTemplate> {
 
-  @NonNull
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/AspectWithExtraInfo.java:15]: What's the motivation behind this change? Is javax's Nonnull different from lombox's NonNull annotation?

Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/AspectWithExtraInfo.java]: Didn't intend to push this change, thanks for catching it.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 8]
```diff
@@ -281,10 +281,10 @@ public boolean isLocalSecondaryIndexEnabled() {
    * @param urn the URN for the entity the aspect is attached to
    * @param auditStamp the audit stamp for the operation
    * @param updateLambda a lambda expression that takes the previous version of aspect and returns the new version
-   * @return {@link RecordTemplate} of the new value of aspect
+   * @return optional {@link RecordTemplate} of the new value of aspect
    */
   @Nonnull
-  public <ASPECT extends RecordTemplate> ASPECT add(@Nonnull URN urn, @Nonnull Class<ASPECT> aspectClass,
+  public <ASPECT extends RecordTemplate> Optional<ASPECT> add(@Nonnull URN urn, @Nonnull Class<ASPECT> aspectClass,
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:287]: I understand the need to support adding a null `newValue` to accommodate delete use case but this could be done without changing the signature for public add method. I don't think it's a good idea to allow users add a null new value for the aspect, writing null newValue should only be explicit to delete.

Maybe just move the common logic to a private add method and keep the public add untouched. WDYT?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (keremsahin1) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java:74]: +1 to Naga, I think explicit is better.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 8 above)
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: Thanks for the suggestion. I have made the change where the `add` method/interface stays the same and only supports adding non-null metadata. The `delete` method will only support adding `null` metadata.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanMetadataAspect.java]: Sure. Made the change.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 9]
```diff
@@ -271,6 +271,58 @@ public boolean isLocalSecondaryIndexEnabled() {
...
+    }
+
+    if (newValue != null && _modelValidationOnWrite) {
+      validateAgainstSchema(newValue);
+    }
+
+    // 2. Invoke pre-update hooks, if any, only if the new metadata being added is not null
+    if (newValue != null && _aspectPreUpdateHooksMap.containsKey(aspectClass)) {
+      _aspectPreUpdateHooksMap.get(aspectClass).forEach(hook -> hook.accept(urn, newValue));
+    }
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: looks like a race condition. I did some more refactor where these are only added to the `add` method since `newValue != null` for the same.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 10]
```diff
@@ -271,6 +271,58 @@ public boolean isLocalSecondaryIndexEnabled() {
...
+
+    if (newValue != null) {
+      checkValidAspect(newValue.getClass());
+    }
+
+    if (newValue != null && _modelValidationOnWrite) {
+      validateAgainstSchema(newValue);
+    }
+
+    // 2. Invoke pre-update hooks, if any, only if the new metadata being added is not null
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: Let me get rid of the numbering completely.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 11]
```diff
@@ -352,6 +378,30 @@ public boolean isLocalSecondaryIndexEnabled() {
...
+    checkValidAspect(aspectClass);
+
+    runInTransactionWithRetry(() -> {
+      final AspectEntry<ASPECT> latest = getLatest(urn, aspectClass);
+      final ASPECT oldValue = latest == null ? null : latest.getAspect();
+
+      return addCommon(urn, latest, oldValue, null, aspectClass, auditStamp, new DefaultEqualityTester<>());
+    }, maxTransactionRetry);
+
+    // 7. TODO: add support for sending MAE for soft deleted aspects
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: So PUH i.e. Post-update hook takes `List<BiConsumer<Urn, RecordTemplate>>`. If you look at the BiConsumer, it takes `RecordTemplate` and not `Optional<RecordTemplate>`. If we really want to support passing `null` metadata to PUH then we should change the signature of the `BiConsumer`. Thoughts?

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 12]
```diff
@@ -591,8 +650,8 @@ public boolean isLocalSecondaryIndexEnabled() {
    * @param version the version for the aspect
    * @param insert use insert, instead of update, operation to save
    */
-  protected abstract void save(@Nonnull URN urn, @Nonnull RecordTemplate value, @Nonnull AuditStamp auditStamp,
-      long version, boolean insert);
+  protected abstract <ASPECT extends RecordTemplate> void save(@Nonnull URN urn, @Nonnull RecordTemplate value,
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:651]: Sure

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 13]
```diff
@@ -352,6 +378,30 @@ public boolean isLocalSecondaryIndexEnabled() {
...
+  /**
+   * Deletes the latest version of aspect for an entity.
+   *
+   * <p>The new aspect will have an automatically assigned version number, which is guaranteed to be positive and
+   * monotonically increasing. Older versions of aspect will be purged automatically based on the retention setting.
+   *
+   * @param urn the URN for the entity the aspect is attached to
+   * @param auditStamp the audit stamp for the operation
+   */
+  private <ASPECT extends RecordTemplate> void delete(@Nonnull URN urn, @Nonnull Class<ASPECT> aspectClass,
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:388]: Good catch, let me fix it.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (keremsahin1) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: Thanks, I see the challenge. I'm perfectly fine to keep that out of scope. Let's add a note to capture that we don't support PUH on delete.

[the remark this answers is not in the document]
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: Sure.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 14]
```diff
@@ -271,6 +271,45 @@ public boolean isLocalSecondaryIndexEnabled() {
...
 
+  /**
+   * Logic common to both {@link #add(Urn, Class, Function, AuditStamp)} and {@link #delete(Urn, Class, AuditStamp, int)} methods.
+   *
+   * <p>The metadata being added is null for the delete method and non-null for the add method.
+   *
+   * @param urn the URN for the entity the aspect is attached to
+   * @param oldValue old metadata of the aspect
+   * @param newValue new metadata of the aspect. This is null for the delete method/operation
+   * @param auditStamp the audit stamp for the operation
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:282]: Missing some javadocs here

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 15]
```diff
@@ -587,12 +646,13 @@ public boolean isLocalSecondaryIndexEnabled() {
...
    * @param value the aspect to save
+   * @param aspectClass the type of aspect to save
    * @param auditStamp the {@link AuditStamp} for the aspect
    * @param version the version for the aspect
    * @param insert use insert, instead of update, operation to save
    */
-  protected abstract void save(@Nonnull URN urn, @Nonnull RecordTemplate value, @Nonnull AuditStamp auditStamp,
-      long version, boolean insert);
+  protected abstract <ASPECT extends RecordTemplate> void save(@Nonnull URN urn, @Nonnull RecordTemplate value,
+      @Nonnull Class<ASPECT> aspectClass, @Nonnull AuditStamp auditStamp, long version, boolean insert);
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:655]: Is value supposed to be nullable here?

Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:655]: Correct, I made the changes in `EbeanLocalDAO` but not here. Thanks for pointing

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 14 above)
Comment (jywadhwani) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: Added

**Review by nagarjunakanamarlapudi — APPROVED** [review]
LGTM

**Review by keremsahin1 — APPROVED** [review]
LGTM, thanks for addressing all the comments!
